packages/block-editor/src/utils/block-removal.js

```diff
@@ -0,0 +1,89 @@
+/**
+ * WordPress dependencies
+ */
+import { useEffect } from '@wordpress/element';
+import { useDispatch, useSelect } from '@wordpress/data';
+import {
+	Modal,
+	Button,
+	__experimentalHStack as HStack,
+} from '@wordpress/components';
+import { __, _n } from '@wordpress/i18n';
+
+/**
+ * Internal dependencies
+ */
+import { store as blockEditorStore } from '../store';
+
+export const blockTypePromptMessages = {
+	'core/query': __( 'Query Loop displays a list of posts.' ),
+	'core/post-content': __(
+		'Post Content displays the content of a post or page.'
+	),
+};
+
+export function BlockRemovalWarningModal() {
+	const { removalFunction, blocksToPromptFor: blockNames } = useSelect(
+		( select ) => select( blockEditorStore ).isRemovalPromptDisplayed()
+	);
+
+	const { displayRemovalPrompt, removalPromptExists } =
+		useDispatch( blockEditorStore );
+
+	// Signalling the removal prompt is in place.
+	useEffect( () => {
+		removalPromptExists( true );
+		return () => {
+			removalPromptExists( false );
+		};
+	}, [ removalPromptExists ] );
```

Comment:
Why is this performed in a `useEffect` when the modal is rendered? Would it be better to set this status when the editor is initialized, e.g. here: https://github.com/WordPress/gutenberg/blob/aba94155dcad8202a4210ba5c2d31ce172bf25aa/packages/edit-site/src/index.js#L38

Reply:
Good point 👀

Reply:
I gave this a go in a5cce0a9ed18ab72c5f37a2df5b355eb7cfa53ad but decided against per my commit message. WDYT?

Reply:
The argument in your commit message sounds reasonable to me! I don't feel strongly about it, and you make a good point that keeping it in the `useEffect` means that the modal component is kind of encapsulated in that it's responsible for handling its own state and setup. Sounds good to leave as-is for now 👍